src/MongoDB/TopologyDescription.c

```diff
@@ -0,0 +1,248 @@
+#include "php_phongo.h"
+
+zend_class_entry* php_phongo_topologydescription_ce;
+
+#define PHONGO_TOPOLOGY_UNKNOWN "Unknown"
+#define PHONGO_TOPOLOGY_SINGLE "Single"
+#define PHONGO_TOPOLOGY_SHARDED "Sharded"
+#define PHONGO_TOPOLOGY_REPLICA_SET_NO_PRIMARY "ReplicaSetNoPrimary"
+#define PHONGO_TOPOLOGY_REPLICA_SET_WITH_PRIMARY "ReplicaSetWithPrimary"
+
+/* {{{ proto array MongoDB\Driver\TopologyDescription::getServers()
+   Returns an array of ServerDescription objects for all known servers in the topology */
+static PHP_METHOD(TopologyDescription, getServers)
+{
+	php_phongo_topologydescription_t* intern;
+	mongoc_server_description_t**     sds;
+	size_t                            i, n = 0;
+
+	intern = Z_TOPOLOGYDESCRIPTION_OBJ_P(getThis());
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	sds = mongoc_topology_description_get_servers(intern->topology_description, &n);
+
+	array_init_size(return_value, n);
+
+	for (i = 0; i < n; i++) {
+		zval obj;
+
+		phongo_serverdescription_init(&obj, sds[i]);
+		add_next_index_zval(return_value, &obj);
+	}
+} /* }}} */
+
+/* {{{ proto boolean MongoDB\Driver\TopologyDescription::hasReadableServer([?MongoDB\Driver\ReadPreference $readPreference])
+    Returns whether the topology has a readable server available */
+static PHP_METHOD(TopologyDescription, hasReadableServer)
+{
+	php_phongo_topologydescription_t* intern;
+	const mongoc_read_prefs_t*        read_preference   = NULL;
+	zval*                             z_read_preference = NULL;
+
+	intern = Z_TOPOLOGYDESCRIPTION_OBJ_P(getThis());
+
+	PHONGO_PARSE_PARAMETERS_START(0, 1)
+	Z_PARAM_OPTIONAL
+	Z_PARAM_OBJECT_OF_CLASS(z_read_preference, php_phongo_readpreference_ce)
+	PHONGO_PARSE_PARAMETERS_END();
+
+	if (z_read_preference) {
+		read_preference = phongo_read_preference_from_zval(z_read_preference);
+	}
+
+	RETVAL_BOOL(mongoc_topology_description_has_readable_server(intern->topology_description, read_preference));
+} /* }}} */
+
+/* {{{ proto boolean MongoDB\Driver\TopologyDescription::hasWritableServer()
+    Returns whether the topology has a writable server available */
+static PHP_METHOD(TopologyDescription, hasWritableServer)
+{
+	php_phongo_topologydescription_t* intern = Z_TOPOLOGYDESCRIPTION_OBJ_P(getThis());
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	RETVAL_BOOL(mongoc_topology_description_has_writable_server(intern->topology_description));
+} /* }}} */
+
+/* {{{ proto string MongoDB\Driver\TopologyDescription::getType()
+    Returns the topology type */
+static PHP_METHOD(TopologyDescription, getType)
+{
+	php_phongo_topologydescription_t* intern = Z_TOPOLOGYDESCRIPTION_OBJ_P(getThis());
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	RETVAL_STRING(mongoc_topology_description_type(intern->topology_description));
+} /* }}} */
+
+/* {{{ MongoDB\Driver\TopologyDescription function entries */
+/* clang-format off */
+ZEND_BEGIN_ARG_INFO_EX(ai_TopologyDescription_hasReadableServer, 0, 0, 0)
+	ZEND_ARG_OBJ_INFO(0, readPreference, MongoDB\\Driver\\ReadPreference, 1)
```

Comment:
I _think_ we should specify `0` for the final macro argument here (`allow_null`). I realize this contradicts what I said previously in https://github.com/mongodb/mongo-php-driver/pull/1244#issuecomment-896114319, so please test it locally to be sure. Try to run what you have now in a test that passes `null` for ReadPreference and see if it fails. If it does, that would make for a reasonable "error" test for the method (along with passing some other bad types, like an integer or an empty array). We have a few such tests in the driver to check for type errors on various methods, so you can base it on that.

Looking at what you have in the `hasReadableServer` implementation earlier in this file, I don't actually think `null` will be accepted as it'd fail the `Z_PARAM_OBJECT_OF_CLASS` check. `Z_PARAM_OPTIONAL` is independent of the parameter's _value_ -- it's just controlling whether we can omit the parameter entirely. In the old `zend_parse_paramaters` syntax, `|` and `!` modifiers were used for optional and nullable, respectively. Offhand, I don't know what the modern equivalent of `!` is (I don't believe it's covered in [Registering and using PHP functions](https://www.phpinternalsbook.com/php7/extensions_design/php_functions.html either)).

Reply:
Just checked and what you said seems right, so I'll add the error test for `hasReadableServer`.